File: frontend/packages/gateway/api/src/activity.ts
import type { DeclinedReason, RefusalKind, SurfacedFinding } from './refusal';
import { apiUrl, askedForJson } from './request';

/** Where a fill of one folder stands. */
export type FillStatus =
  /** Armed, or walking the folder. */
  | 'filling'
  /** Every file it set out to bring over is here or accounted for. */
  | 'done'
  /** It stopped short, and `stopped` says what stopped it. */
  | 'stopped'
  /** A fetch landed in another folder and the fill followed it there. */
  | 'superseded';

/**
 * One refusal, in the shape every refusal from this server takes.
 *
 * The same fields a refused request carries, so a declined Entry is read with
 * the same three branches: which kind, which way, which finding.
 */
export interface Refused {
  error: RefusalKind;
  message: string;
  reason?: DeclinedReason;
  surfaced?: SurfacedFinding;
}

/** One Entry a fill did not bring over, and what it found instead. */
export interface DeclinedEntry extends Refused {
  path: string;
}

/**
 * Which phase of a flow a run is in.
 *
 * The whole set is named here for the reason every other union in this package
 * is: a caller writes a branch per phase, and one it has never heard of is one
 * it falls off the end of.
 *
 * They are the use case layer's own phases rather than this client's reading of
 * them — the same ones the command line draws its progress line from — so a
 * browser and a terminal watching one run cannot call it two different things.
 */
export type Phase =
  /** Bringing this device's catalog up to the Library's head. */
  | 'catching_up'
  /** Settling the pending rows an interrupted run left behind. */
  | 'reconciling'
  /** Reading this device's mapped folders and deciding what the run will do. */
  | 'scanning'
  /** Encoding local files into Containers on this device. */
  | 'packing'
  /** Sending encoded Containers to Storage. */
  | 'uploading'
  /** Reading Containers back from Storage and placing the files in them. */
  | 'fetching';

/**
 * How far into one phase a run has got.
 *
 * A count of things done out of things to do, and never a byte count: what a
 * person watching a transfer wants to know is whether it is moving and roughly
 * how much is left.
 */
export interface Step {
  phase: Phase;
  /** How many units of that phase have finished. */
  done: number;
  /**
   * How many there are in all, and `null` where the phase cannot say.
   *
   * Not the same state as `0`, and they must not be shown alike: a phase that
   * cannot count its work is exactly the phase that goes quiet for minutes, and
   * what a screen shows for it is the phase's name without numbers.
   */
  total: number | null;
}

/**
 * What the server is bringing over on its own.
 *
 * Opening a file this device does not have fetches it and then goes on to fetch
 * the rest of its folder, unasked — nobody who opened page one stops there.
 * This is that work's account of itself, and it is the server's own state: it
 * says nothing about what the Library holds, it is gone when the server is, and
 * `present` and `remote` stay the listing's to say.
 */
export interface Fill {
  /**
   * Which run of the fill this is, counted from the start of the server.
   *
   * What tells one run's account of itself from the next's. A screen that lets
   * somebody put a line away remembers which run they put away by this, so that
   * doing so does not silently take the next run's line with it.
   */
  run: number;
  /** The folder being brought over; the Library root is the empty string. */
  folder: string;
  status: FillStatus;
  /**
   * How many of the folder's files the fill set out to bring over, and `0`
   * until it has read the folder's listing.
   */
  total: number;
  /** How many of them are on this device now. */
  done: number;
  /** The Entries it declined, each with what opening it would have said. */
  declined: DeclinedEntry[];
  /**
   * The folders asked for by name that are waiting their turn behind this one,
   * oldest first.
   *
   * The names and not a count, for the reason a freeze's queue carries names: a
   * person who pressed a button wants to know that theirs is coming. A folder
   * asked for by name waits rather than displacing the fill in progress, and
   * between the press and the run this is the only thing said about it — the
   * line names the folder being brought over, and the button that named this
   * one goes away as the queue takes it.
   */
  waiting: string[];
  /**
   * The folders that were queued behind a fill and thrown away when the work
   * ended without an answer, and that nobody has taken up since.
   *
   * Not this fill's own doing and not gone when it is: the line and the retry
   * beside it name the folder that died, and these are the ones nothing on the
   * screen would otherwise mention at all. Asking for one again is what takes
   * it off this list.
   */
  dropped: string[];
  /**
   * The runs that stopped and that a later one took the record from, oldest
   * first.
   *
   * Not the list above: those folders were thrown away before anything started
   * on them, and these ran and got part way. What the server holds for a browser
   * is one run at a time, and the next folder off the queue takes the record
   * from a stopped one within a tick — somebody clicking into another folder
   * while Storage is down is enough — so without these the line naming that
   * folder, the Entries it declined and the offer of a second attempt would all
   * go unread.
   *
   * Each says what its own run came to and nothing about the queue: `waiting`,
   * `dropped` and this list belong to the flow rather than to any run, so they
   * are reported once, on the run the flow is on, and arrive empty here.
   */
  displaced: Fill[];
  /** What stopped the fill, and `null` where nothing did. */
  stopped: Refused | null;
}

/** Where a sync stands. */
export type SyncStatus =
  /** Armed, or walking the mapped folders. */
  | 'syncing'
  /** It finished, whatever it found. */
  | 'done'
  /** It stopped short, and `stopped` says what stopped it. */
  | 'stopped';

/**
 * One thing a sync that succeeded still has to say.
 *
 * Not a refusal: nothing was refused, the run succeeded, and this is what it
 * left alone — a file whose Entry lives in a Pack, a file this device no longer
 * has, a mapped root it could not vouch for. Reading only the counts would tell
 * somebody their file is backed up when it is not.
 */
export interface SyncFinding {
  /** The Entry this is about, and `null` where it is about no single one. */
  path: string | null;
  message: string;
}

/**
 * What the server is carrying into the Library on its own.
 *
 * Dropping a file means "add this", and adding is not finished when the bytes
 * reach the folder: the server runs the same sync the person would have typed,
 * and this is that run's account of itself. Like a fill it is the server's own
 * state — gone when the server is, and never uploaded.
 */
export interface Sync {
  /** Which run of the sync this is, counted from the start of the server. */
  run: number;
  status: SyncStatus;
  /** How many files the run carried in, and `0` until it is over. */
  added: number;
  /** What it found and did not act on. */
  noted: SyncFinding[];
  /**
   * How far into the walk the flow says it has got, and `null` before it has
   * said and once the run is over.
   */
  step: Step | null;
  /** What stopped the sync, and `null` where nothing did. */
  stopped: Refused | null;
}

/** Where a freeze of one folder stands. */
export type FreezeStatus =
  /** Armed, or packing the folder. */
  | 'freezing'
  /** It finished, whatever it found. */
  | 'done'
  /** It stopped short, and `stopped` says what stopped it. */
  | 'stopped';

/**
 * What the server is packing into the Library on its own.
 *
 * Dropping a book into a folder made for it means "bring this in", and a book is
 * the one thing a sync is the wrong shape for: a folder of a few hundred page
 * images would become a few hundred Storage objects, a few hundred uploads, and
 * a few hundred calls to open it again. So the server packs them instead, and
 * this is that run's account of itself. Like a fill and a sync it is the
 * server's own state — gone when the server is, and never uploaded.
 *
 * The counts are outcomes and stay `0` until the batch commits: a freeze builds
 * and commits one batch, so until it has committed no number of Packs would be
 * true. Where the run has got to is a different question and `step` answers it.
 */
export interface Freeze {
  /** Which run of the freeze this is, counted from the start of the server. */
  run: number;
  /** The folder being packed; the Library root is the empty string. */
  folder: string;
  status: FreezeStatus;
  /** How many Packs the run built, and `0` until it is over. */
  packs: number;
  /** How many Entries those Packs hold, and `0` until it is over. */
  entries: number;
  /** What it found and did not act on. */
  noted: SyncFinding[];
  /**
   * How far into the run the flow says it has got, and `null` before it has
   * said and once the run is over.
   */
  step: Step | null;
  /**
   * The books waiting their turn behind this one, oldest first.
   *
   * The names and not a count: a person who dropped a second book wants to know
   * that theirs is queued, and a bare number says only that somebody's is.
   */
  waiting: string[];
  /** The books thrown away when the work ended without an answer. */
  dropped: string[];
  /**
   * The runs that stopped and that a later one took the record from, oldest
   * first.
   *
   * Not the list above: those books were thrown away before anything started on
   * them, and these ran and got part way. A freeze Storage stopped leaves its
   * pages on the disk and out of the Library, in a folder made in this browser
   * that the Library has never heard of — so this run is the only thing naming
   * that place, and a second book queued behind the first is all it takes for
   * the next run to take the record from it. Without these a reload would draw
   * no row for the folder, offer no way to walk in, and make no second attempt
   * at it.
   *
   * Each says what its own run came to and nothing about the queue, on the terms
   * a fill's do.
   */
  displaced: Freeze[];
  /** What stopped the freeze, and `null` where nothing did. */
  stopped: Refused | null;
}

/**
 * Which of the two states this device holds the Library in.
 *
 * The two words the spec calls them by, and the only two there are: the
 * Passphrase moves a device from the first to the second, and a lock — one
 * somebody asked for, or the interval the server went unasked for — moves it
 * back.
 */
export type LibraryState = 'locked' | 'unlocked';

/**
 * How far this device's catalog has got with the Library.
 *
 * Every listing comes out of the catalog, and the catalog holds what this
 * device has replayed — so an explorer showing nothing is showing either a
 * Library with nothing in it or a device that has not learnt what is in it, and
 * from the rows alone the two are the same screen.
 */
export type CatalogState =
  /** A catch-up is running; what is listed is what was known before it began. */
  | 'catching_up'
  /** The last one finished, so this device has replayed what the Library had. */
  | 'caught_up'
  /** The last one did not finish, and `trouble` says what stopped it. */
  | 'behind';

/** How the catalog stands, and what stopped it where something did. */
export interface Catalog {
  state: CatalogState;
  /** What stopped the last catch-up, and `null` where nothing did. */
  trouble: Refused | null;
}

/** What the server is doing on its own — `GET /api/activity`. */
export interface Activity {
  /**
   * What the process that answered calls itself.
   *
   * The same string for every answer one process gives, and a different one
   * after it is started again. It says nothing about the device or the
   * Library — the server composes it out of entropy as it starts — and what it
   * is for is everything a page remembers between answers that is true of one
   * process only: the run numbers below all start again at 1 with the next one,
   * so a name the page has not seen before is the sign to forget what it was
   * holding. The run numbers cannot say it themselves, since one lower than a
   * run already put away is equally what an answer issued just before the
   * dismissal looks like.
   *
   * It matters because a restart is ordinary rather than exceptional: a locked
   * Library is unlocked by typing the Passphrase and starting the server again,
   * and a tab left open across one would otherwise hide the new process's first
   * runs — a fill's line and its declined Entries, and the one sentence saying
   * a sync did not back a file up.
   */
  server: string;
  /**
   * Whether this device still holds the Library open.
   *
   * Read from this route and from no other, because this is the one a screen
   * asks without being told to. A lock that happened on the server's own clock
   * tells nobody, so a window left open over a page it decrypted learns of it
   * here or not until its next request is refused.
   */
  library: LibraryState;
  /**
   * How this device's catalog stands with the Library.
   *
   * Here for the reason the field above it is: it is what the server did to
   * itself rather than work anybody asked for, and this is the one question a
   * page asks without being told to — which is exactly when an empty Library
   * has to be told apart from a Library nobody has looked at yet.
   */
  catalog: Catalog;
  /** The latest fill, running or finished, and `null` where none has run. */
  fill: Fill | null;
  /** The latest sync, running or finished, and `null` where none has run. */
  sync: Sync | null;
  /** The latest freeze, running or finished, and `null` where none has run. */
  freeze: Freeze | null;
}

/** Asks what the server is doing on its own. */
export function getActivity(signal?: AbortSignal): Promise<Activity> {
  return askedForJson<Activity>(apiUrl('activity'), signal);
}

/**
 * Carries the mapped folders into the Library again — `POST /api/sync`.
 *
 * Not a "sync now" button and not offered as one. What syncs a dropped file is
 * dropping it; this exists for the state that leaves behind — a sync Storage
 * stopped, whose files are sitting in the folder with nothing left to drop —
 * where the alternative is telling somebody to add a file they have added.
 *
 * It takes no folder. Which folders a sync walks is the device's mappings and
 * never an argument, here as on the command line.
 */
export function startSync(signal?: AbortSignal): Promise<Activity> {
  return askedForJson<Activity>(apiUrl('sync'), signal, 'POST');
}

/**
 * Takes one folder up again — `POST /api/fill?path=`.
 *
 * Not a download button and not offered as one. What brings a folder over is
 * opening a file in it; this exists for the three states that leaves behind — a
 * fill Storage stopped, one superseded when somebody clicked elsewhere, and a
 * folder a worker that died threw away before it began — where the alternative
 * is telling a person to open a file they have opened.
 *
 * What it asks for waits its turn rather than displacing the fill in progress,
 * unlike the fetch that arms one implicitly: every folder named here was named
 * by a button somebody pressed, so two presses bring both folders over, in the
 * order they were pressed.
 *
 * It answers with the activity as it stands the moment the fill is armed rather
 * than waiting for the work, which is why the caller goes on polling.
 */
export function startFill(folder: string, signal?: AbortSignal): Promise<Activity> {
  return askedForJson<Activity>(
    apiUrl('fill', folder === '' ? undefined : { path: folder }),
    signal,
    'POST',
  );
}

/**
 * Packs one folder into Packs again — `POST /api/freeze?path=`.
 *
 * Not a "pack this" button and not offered as one. What packs a book is bringing
 * it in — dropping its pages onto a folder made a moment ago, which arms this
 * itself — and this exists for the state that leaves behind: a freeze Storage
 * stopped, whose pages are sitting in the folder with nothing left to drop,
 * where the alternative is telling somebody to drop a book they have dropped.
 *
 * It takes a folder, unlike the sync: a freeze is of one folder, and one
 * narrowed to nothing would pack the whole Library.
 *
 * It answers with the activity as it stands the moment the freeze is armed
 * rather than waiting for the work, which is why the caller goes on polling.
 */
export function startFreeze(folder: string, signal?: AbortSignal): Promise<Activity> {
  return askedForJson<Activity>(
    apiUrl('freeze', folder === '' ? undefined : { path: folder }),
    signal,
    'POST',
  );
}

File: backend/crates/apps/coffret-server/src/routes/activity.rs
use std::sync::Arc;

use axum::extract::State;
use axum::Json;
use serde::Serialize;

use coffret_device::{Phase, Step};

use crate::fill::{Activity, Declined};
use crate::finding::Finding;
use crate::folder::Folder;
use crate::freeze::FreezeActivity;
use crate::latest::Latest;
use crate::refresh::Standing;
use crate::reported::Reported;
use crate::state::ServerState;
use crate::sync::SyncActivity;

/// What the server is doing on its own, which is three things — and the one
/// thing it may have done to itself.
///
/// A fill, a sync and a freeze. Everything else this server does it does because
/// a request asked it to, and a request is answered rather than reported on;
/// these three are the work nobody asked for — the rest of a folder being
/// brought over behind a reader, files somebody dropped being carried into the
/// Library, and a book somebody brought in being packed — so they are what there
/// is to tell a browser about.
///
/// Side by side and not one after another: they are separate work over one
/// Library, any of them can be running without the others, and a browser reads
/// each on its own.
///
/// The first two fields are not work. Whether this device still holds the
/// Library open is the one thing about itself a browser cannot find out by
/// waiting: the idle lock happens on this server's clock and tells nobody
/// (spec: DK-4), so a window left open over a page it decrypted would go on
/// showing that page until something it asked for happened to be refused. It
/// rides here because this is the question an open explorer is already asking,
/// and because asking it keeps nothing awake.
///
/// How the catalog stands rides here for the same reason and answers the same
/// kind of question. Every listing comes out of the catalog, and the catalog is
/// what this device has replayed (spec: CK-9) — so a device fresh from `join`
/// whose catch-up did not land serves an empty Library that is not empty, and no
/// request a screen makes for its own sake would ever say so. This is the one a
/// page asks as it comes up.
///
/// And the first field of all is not about the Library either: it is which
/// process answered. Everything below it that a browser remembers between
/// answers — which run's line somebody read and put away — is true of one
/// process only, because the run numbers start again at 1 with the next one.
#[derive(Serialize)]
pub struct ActivityDto {
    /// What the process that answered calls itself.
    ///
    /// The same string for every answer this process gives and a different one
    /// after it is started again, which is all a page needs from it: what it
    /// remembers about runs it has been shown is about this server's runs, and
    /// a name it has not seen before is the sign to forget it. A page cannot
    /// reach that conclusion from the run numbers — one lower than a run it put
    /// away is equally what an answer issued just before the dismissal looks
    /// like.
    ///
    /// It says nothing about this device or this Library: see
    /// [`ServerId`](crate::server_id::ServerId) for what it is composed of and
    /// why it is composed rather than taken from something that was already
    /// there.
    server: String,
    /// Which of the two states this device holds the Library in, in the words
    /// DK-1 uses: `locked` or `unlocked`.
    library: &'static str,
    /// How this device's catalog stands with the Library.
    catalog: CatalogDto,
    /// The latest fill, running or finished, and `null` where none has run.
    fill: Option<FillDto>,
    /// The latest sync, running or finished, and `null` where none has run.
    sync: Option<SyncDto>,
    /// The latest freeze, running or finished, and `null` where none has run.
    freeze: Option<FreezeDto>,
}

/// How far this device has got with the Library, and what stopped it where
/// something did.
///
/// Two fields and not one word, because a person is owed the reason as well as
/// the state: "this device has not caught up" is something to wait through, and
/// "Storage did not answer" is something to press a button about.
#[derive(Serialize)]
struct CatalogDto {
    /// `catching_up`, `caught_up` or `behind`.
    state: &'static str,
    /// What stopped the last catch-up, and `null` where nothing did.
    trouble: Option<RefusalDto>,
}

/// How far into a run the flow reports it has got.
///
/// The [`Step`](coffret_device::Step) the use case layer reports, in the shape a
/// browser reads it in. It is the same value the command line renders its
/// progress line from, and deliberately so: one run has one answer about where
/// it is, whichever shell is watching.
#[derive(Serialize)]
struct StepDto {
    /// Which phase of the flow it is in.
    phase: &'static str,
    /// How many units of that phase are finished.
    done: usize,
    /// How many there are in all, and `null` where the phase cannot say.
    ///
    /// Absent rather than zero, and the two must not be shown alike: a phase
    /// that cannot count its work — a catch-up learns what it has to replay by
    /// replaying it — is exactly the phase that goes quiet for minutes, and a
    /// `0` there would read as a phase with nothing in it.
    total: Option<usize>,
}

#[derive(Serialize)]
struct FillDto {
    /// Which run of the fill this is, counted from the start of this process.
    ///
    /// What tells one run's account of itself from the next's, so that a line
    /// somebody has read and put away does not take the next run's with it.
    run: u64,
    /// The folder being brought over; the Library root is the empty string, as
    /// it is in a listing.
    folder: String,
    /// `filling`, `done`, `stopped` or `superseded`.
    status: &'static str,
    /// How many of the folder's files the fill set out to bring over, and `0`
    /// until it has read the folder's listing.
    total: usize,
    /// How many of them are on this device now.
    done: usize,
    /// The Entries it did not bring over, each with what the file route would
    /// have said about it — so a row can be marked without anyone clicking it.
    declined: Vec<DeclinedDto>,
    /// The folders somebody asked for by name that are waiting their turn behind
    /// this one, oldest first.
    ///
    /// The names and not the length, for the reason the freeze's queue carries
    /// names: a person who pressed a button wants to know that *their* folder is
    /// coming. A folder asked for by name waits rather than displacing the fill
    /// in progress, and between the press and the run this is the only thing on
    /// the wire about it — the line names the folder being brought over, and the
    /// button that named this one is gone the moment the queue takes it.
    waiting: Vec<String>,
    /// The folders that were queued behind a fill and thrown away when the work
    /// ended without an answer, and that nobody has taken up since.
    ///
    /// Not this fill's own doing and not gone when it is: it is the queue that
    /// lost them, and what settles one is somebody asking for that folder again.
    /// Without it the line and the retry both name the folder that died, and the
    /// folder somebody clicked into afterwards is never mentioned at all.
    dropped: Vec<String>,
    /// The runs that stopped and that a later one took the record from, oldest
    /// first — the newest eight at most, the oldest forgotten past that (see
    /// the fill's `Progress`).
    ///
    /// A fill Storage stopped is a folder somebody asked for and did not get,
    /// and the field above this one is not what it is: those folders were thrown
    /// away before anything started on them, and these ran and got part way. The
    /// next folder taken off the queue takes the record from a stopped run
    /// within a tick — a person clicking into another folder while Storage is
    /// down is enough — so without these the line, the Entries it declined and
    /// the offer of a second attempt would all go unread.
    ///
    /// Each carries what its own run came to and nothing of the flow: the three
    /// lists above belong to the queue rather than to any run, so they are the
    /// run on record's to report and arrive empty here.
    displaced: Vec<FillDto>,
    /// What stopped the fill, and `null` where nothing did.
    stopped: Option<RefusalDto>,
}

#[derive(Serialize)]
struct SyncDto {
    /// Which run of the sync this is, counted from the start of this process.
    run: u64,
    /// `syncing`, `done` or `stopped`.
    status: &'static str,
    /// How many files the run carried into the Library, and `0` until it is
    /// over.
    added: usize,
    /// What the run found and did not act on — a file inside a Pack it cannot
    /// replace, a file this device no longer has, a mapped root the device
    /// could not vouch for.
    findings: Vec<FindingDto>,
    /// How far into the walk the flow says it has got, and `null` before it has
    /// said and once the run is over.
    step: Option<StepDto>,
    /// What stopped the sync, and `null` where nothing did.
    stopped: Option<RefusalDto>,
}

#[derive(Serialize)]
struct FreezeDto {
    /// Which run of the freeze this is, counted from the start of this process.
    run: u64,
    /// The folder being packed; the Library root is the empty string, as it is
    /// in a listing.
    folder: String,
    /// `freezing`, `done` or `stopped`.
    status: &'static str,
    /// How many Packs the run built, and `0` until it is over.
    packs: usize,
    /// How many Entries those Packs hold, and `0` until it is over.
    entries: usize,
    /// What the run found and did not act on — a page inside a Pack it cannot
    /// replace, a mapped root the device could not vouch for. A page a Pack
    /// holds and that did not change is not among these: it is not eligible in
    /// the first place (spec: PK-1), and a second run over a book saying so of
    /// every page would be a wall of findings about nothing.
    findings: Vec<FindingDto>,
    /// How far into the run the flow says it has got, and `null` before it has
    /// said and once the run is over.
    ///
    /// The one thing a person dropping several hundred pages had no way to see:
    /// the counts above are outcomes and stay `0` until the batch commits, and
    /// this is the run moving while it builds it.
    step: Option<StepDto>,
    /// The books waiting their turn behind this one, oldest first.
    ///
    /// The names and not the length, because the length is the least of it: a
    /// person who dropped a second book wants to know that *their* book is
    /// queued, and a bare `1` says only that somebody's is. A freeze commits one
    /// batch (spec: PK-7), so a second one waits rather than displacing the
    /// first, and until it starts there is nothing else on the screen about it.
    waiting: Vec<String>,
    /// The books thrown away when the work ended without an answer, and that
    /// nobody has asked for since.
    dropped: Vec<String>,
    /// The runs that stopped and that a later one took the record from, oldest
    /// first.
    ///
    /// A freeze Storage stopped leaves a folder of pages on the disk and out of
    /// the Library, in a folder the browser made and the Library has never heard
    /// of — so this run is the only thing on the wire naming that place. The next
    /// book off the queue takes the record from it, and a second book queued
    /// behind the first is all it takes: without these the folder would be in
    /// none of the lists a page reads back, and a reload would draw no row for
    /// it, offer no way to walk into it and make no second attempt at it.
    ///
    /// The field above this one is not what these are: those books were thrown
    /// away before anything started on them, and these ran and got part way.
    ///
    /// Each carries what its own run came to and nothing of the flow: the three
    /// lists above belong to the queue rather than to any run, so they are the
    /// run on record's to report and arrive empty here.
    displaced: Vec<FreezeDto>,
    /// What stopped the freeze, and `null` where nothing did.
    stopped: Option<RefusalDto>,
}

/// One thing a run that succeeded still has to say.
///
/// Unlike a declined Entry this carries no refusal kind, and deliberately:
/// nothing was refused. The run succeeded and left this alone, so what there is
/// to show is the sentence and the row it belongs to — `null` for the findings
/// that are about no single Entry — and which finding it is, named beside the
/// sentence in the `reason` and `surfaced` a declined Entry names the same
/// state by, so that a page branching on it never has to read it out of prose.
#[derive(Serialize)]
struct FindingDto {
    path: Option<String>,
    message: String,
    /// One of the names [`Finding::reason`] lists.
    reason: &'static str,
    /// The device layer's name for a finding about one Entry, and absent for
    /// one about a mapping or a Container — left out rather than `null`, as a
    /// refusal's is, so the two read alike.
    #[serde(skip_serializing_if = "Option::is_none")]
    surfaced: Option<&'static str>,
}

#[derive(Serialize)]
struct DeclinedDto {
    path: String,
    #[serde(flatten)]
    refusal: RefusalDto,
}

/// One refusal, in the shape every refusal on these routes takes.
///
/// The same four fields under the same four names, so a browser reads a
/// declined Entry with the code it already has for a refused request. Shared
/// with the upload's per-part list for that reason: a refusal a person meets by
/// dropping a file and one they meet by opening it are one vocabulary.
#[derive(Serialize)]
pub(super) struct RefusalDto {
    error: &'static str,
    message: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    reason: Option<&'static str>,
    #[serde(skip_serializing_if = "Option::is_none")]
    surfaced: Option<&'static str>,
}

impl ActivityDto {
    /// What the server is doing right now, as the browser is told it.
    ///
    /// Read off the state rather than assembled by each caller: the three routes
    /// that answer with an activity all answer with the whole of it, and a
    /// caller that assembled two thirds of it would be a browser told that
    /// whichever work it did not name had stopped.
    ///
    /// Each of the two flows that keep a queue is read in one go rather than a
    /// field at a time. What a run is, what is waiting behind it and what its
    /// queue lost move together — a worker that dies stops the run and throws
    /// the rest away in one stroke — so three reads could put half of that
    /// change on the wire: a run still filling beside the folders its ending
    /// threw away, or a run still `done` beside a queue that has already been
    /// taken up, which is a browser told there is nothing left to follow at the
    /// moment the next run starts.
    pub fn of(state: &ServerState) -> Self {
        Self {
            server: state.server().to_owned(),
            library: if state.holds_library() {
                "unlocked"
            } else {
                "locked"
            },
            catalog: CatalogDto::of(&state.catalog.standing()),
            fill: state.fills.reported().as_ref().map(FillDto::of),
            sync: state.syncs.activity().as_ref().map(SyncDto::of),
            freeze: state.freezes.reported().as_ref().map(FreezeDto::of),
        }
    }
}

impl CatalogDto {
    fn of(standing: &Standing) -> Self {
        match standing {
            Standing::CatchingUp => Self {
                state: "catching_up",
                trouble: None,
            },
            Standing::CaughtUp => Self {
                state: "caught_up",
                trouble: None,
            },
            Standing::Behind(trouble) => Self {
                state: "behind",
                trouble: Some(RefusalDto::of(trouble)),
            },
        }
    }
}

impl StepDto {
    fn of(step: &Step) -> Self {
        Self {
            phase: named(step.phase),
            done: step.done,
            total: step.total,
        }
    }
}

/// The word one phase travels under.
///
/// Named here rather than on [`Phase`](coffret_device::Phase) because it is this
/// route's vocabulary: the use case layer says what a run is doing, and what a
/// browser calls it is the browser's business. Matched exhaustively, so a phase
/// added to the flow stops this compiling until somebody says what a person
/// reading a status bar should be told it is.
fn named(phase: Phase) -> &'static str {
    match phase {
        Phase::CatchingUp => "catching_up",
        Phase::Reconciling => "reconciling",
        Phase::Scanning => "scanning",
        Phase::Packing => "packing",
        Phase::Uploading => "uploading",
        Phase::Fetching => "fetching",
    }
}

impl SyncDto {
    fn of(activity: &SyncActivity) -> Self {
        Self {
            run: activity.run,
            status: activity.status.as_str(),
            added: activity.added,
            findings: activity.findings.iter().map(FindingDto::of).collect(),
            step: activity.step.as_ref().map(StepDto::of),
            stopped: activity.stopped.as_ref().map(RefusalDto::of),
        }
    }
}

impl FreezeDto {
    /// The whole of what the flow has to say: the run on record, with the queue's
    /// two lists and the runs it took the record from beside it.
    fn of(latest: &Latest<FreezeActivity>) -> Self {
        Self {
            waiting: named_folders(&latest.waiting),
            dropped: named_folders(&latest.dropped),
            displaced: latest.displaced.iter().map(Self::alone).collect(),
            ..Self::alone(&latest.activity)
        }
    }

    /// One run and what it came to, with nothing of the flow around it.
    ///
    /// What a displaced run is answered as, and the half of the run on record
    /// that is about the run rather than about the queue. The queue's lists are
    /// empty here because they are not this run's to report: what is waiting and
    /// what was thrown away belong to the flow, and are said once, on the run the
    /// flow is on.
    fn alone(activity: &FreezeActivity) -> Self {
        Self {
            run: activity.run,
            folder: activity.folder.as_str().to_owned(),
            status: activity.status.as_str(),
            packs: activity.packs,
            entries: activity.entries,
            findings: activity.findings.iter().map(FindingDto::of).collect(),
            step: activity.step.as_ref().map(StepDto::of),
            waiting: Vec::new(),
            dropped: Vec::new(),
            displaced: Vec::new(),
            stopped: activity.stopped.as_ref().map(RefusalDto::of),
        }
    }
}

/// Folders as a listing spells them: the Library root is the empty string.
fn named_folders(folders: &[Folder]) -> Vec<String> {
    folders
        .iter()
        .map(|folder| folder.as_str().to_owned())
        .collect()
}

impl FindingDto {
    fn of(finding: &Finding) -> Self {
        Self {
            path: finding.path.clone(),
            message: finding.message.clone(),
            reason: finding.reason,
            surfaced: finding.surfaced,
        }
    }
}

impl FillDto {
    /// The whole of what the flow has to say: the run on record, with the queue's
    /// two lists and the runs it took the record from beside it.
    fn of(latest: &Latest<Activity>) -> Self {
        Self {
            waiting: named_folders(&latest.waiting),
            dropped: named_folders(&latest.dropped),
            displaced: latest.displaced.iter().map(Self::alone).collect(),
            ..Self::alone(&latest.activity)
        }
    }

    /// One run and what it came to, with nothing of the flow around it.
    ///
    /// What a displaced run is answered as, on the terms a freeze's is: the
    /// queue's lists are not this run's to report.
    fn alone(activity: &Activity) -> Self {
        Self {
            run: activity.run,
            folder: activity.folder.as_str().to_owned(),
            status: activity.status.as_str(),
            total: activity.total,
            done: activity.done,
            declined: activity.declined.iter().map(DeclinedDto::of).collect(),
            waiting: Vec::new(),
            dropped: Vec::new(),
            displaced: Vec::new(),
            stopped: activity.stopped.as_ref().map(RefusalDto::of),
        }
    }
}

impl DeclinedDto {
    fn of(declined: &Declined) -> Self {
        Self {
            path: declined.path.clone(),
            refusal: RefusalDto::of(&declined.refusal),
        }
    }
}

impl RefusalDto {
    pub(super) fn of(refusal: &Reported) -> Self {
        Self {
            error: refusal.kind,
            message: refusal.message.clone(),
            reason: refusal.reason,
            surfaced: refusal.surfaced,
        }
    }
}

/// `GET /api/activity`
///
/// Polled while something is happening and not otherwise: an explorer with
/// nothing in flight asks for nothing. An open reader counts as something
/// happening, which is what carries the lock's news to the one screen holding
/// plaintext.
///
/// It needs no key and takes none, so it answers a locked server as readily as
/// an open one.
pub async fn activity(State(state): State<Arc<ServerState>>) -> Json<ActivityDto> {
    Json(ActivityDto::of(&state))
}

// Every state this answer can be in, written to the file the explorer's own
// cases read back through its types.
#[cfg(test)]
mod contract;
